File: LeetCodePython/py60.py
# 第k个排列，给出集合[1,2,3,...,n]的所有元素的全排列按顺序标号
# 解法：回溯+剪枝；根据排列组合的个数确定每一位数字

class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        nums = [str(i) for i in range(1,n+1)]
        n -= 1
        result = ""
        while n>-1:
            t = math.factorial(n)
            pos = math.ceil(k/t) - 1 # 当k=0时，刚好整除，pos=-1，即逆序输出
            result += nums[pos]
            # print(n,t,pos)
            nums.pop(pos)
            k %= t
            n -= 1
        return result


class Solution2:
    '''
    作者：liweiwei1419
    链接：https://leetcode-cn.com/problems/two-sum/solution/hui-su-jian-zhi-python-dai-ma-java-dai-ma-by-liwei/

    '''
    def getPermutation(self, n: int, k: int) -> str:
        if n == 0:
            return []
        nums = [i + 1 for i in range(n)]
        used = [False for _ in range(n)]

        return self.__dfs(nums, used, n, k, 0, [])

    def __factorial(self, n):
        # 这种编码方式包括了 0 的阶乘是 1 这种情况
        res = 1
        while n:
            res *= n
            n -= 1
        return res

    def __dfs(self, nums, used, n, k, depth, pre):
        if depth == n:
            # 在叶子结点处结算
            return ''.join(pre)
        # 后面的数的全排列的个数
        ps = self.__factorial(n - 1 - depth)
        print(ps)
        for i in range(n):
            # 如果这个数用过，就不再考虑
            if used[i]:
                continue
            # 后面的数的全排列的个数小于 k 的时候，执行剪枝操作
            if ps < k:
                k -= ps
                continue
            pre.append(str(nums[i]))
            # 因为直接走到叶子结点，因此状态不用恢复
            used[i] = True
            return self.__dfs(nums, used, n, k, depth + 1, pre)


        

File: LeetCodeCPlusPlus/py60.cpp
// 第k个排列，集合[1,2,...,n]的全排列按顺序编号
#include <string>
#include <vector>
#include <math.h>
// 解法找规律。算每种排列的个数，得出每一位数字
using namespace std;
class Solution {
public:
    int factorial(int j) {
        if (j == 1 || j == 0) return 1;
        else return j*factorial(j-1);
    }
    string getPermutation(int n, int k) {
        vector<char> nums;
        for (int i=1;i<n+1;i++)
            nums.push_back('0'+i);
        
        n--;
        string result="";
        while (n>-1) {
            int t = factorial(n);
            int pos = ceil(1.0*k/t) - 1;
            //cout<<n<<" "<<t<<" "<<k<<" "<<pos<<endl;
            pos = pos >= 0 ? pos:nums.size()-1;  // 出现-1的情况是k=0，逆序输出，k小于t时，ceil(0.x) = 1
            result.push_back(nums[pos]);
            nums.erase(nums.begin()+pos);
            k %= t;
            n--;
        }
        return result;
        
    }
};